server/fleet/app.go

```diff
@@ -239,6 +258,29 @@ func (c *AppConfig) ApplyDefaults() {
 	c.WebhookSettings.Interval.Duration = 24 * time.Hour
 }
 
+// EnableStrictDecoding enables strict decoding of the AppConfig struct.
+func (c *AppConfig) EnableStrictDecoding() { c.strictDecoding = true }
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (c *AppConfig) UnmarshalJSON(b []byte) error {
+	// Decode any legacy settings that might be present.
+	var legacy legacyConfig
+	if err := json.Unmarshal(b, &legacy); err != nil {
+		return err
+	}
+
+	// TODO(roperzh): define and assign legacy settings to new fields. E.g:
+	// c.Features = legacy.HostSettings
+
+	// Decode the config, overriding any legacy settings.
+	type alias *AppConfig
+	decoder := json.NewDecoder(bytes.NewReader(b))
+	if c.strictDecoding {
+		decoder.DisallowUnknownFields()
+	}
+	return decoder.Decode(alias(c))
```

Comment:
This is the best solution I could come up with that:

1. Hides deprecated fields from `AppConfig`
2. Gives new fields the ability to override deprecated fields

Another alternative I considered is doing something like the untested example below, which does a single `Unmarshal` pass, but gives precedence to deprecated fields over new fields.

```go
type allConfigs struct {
  AppConfig
  legacyFields
}

var aux allConfigs
decoder := json.NewDecoder(bytes.NewReader(b))
if c.strictDecoding {
	decoder.DisallowUnknownFields()
}
if err := decoder.Decode(aux); err != nil {
	return err
}

// I'm writing this in GitHub, not sure if compiles, but you get the idea
c.Features = aux.HostSettings
c = aux.AppConfig
```

I'm also very open to any other ways to solve the main problem.